### ISSUE TITLE
kepler.apk can't be extracted

### ISSUE THREAD

**Issue by vrm42** [sandysound/orvibo-b25-server#1]
Hey, sandysoud! Your solutions looks promising. I hope I can make it work. I got a problem with getting the key from the apk.
The url in the script does not exist. I downloaded the file from other locations, but unzip says that is not a zip file. As I understand that key is not user specific, therefore it is not a sensitive information. Can you share it? Thanks!

**Comment by sandysound** [comment]
Hey, try changing the script to add www to start of orvibo.com/software/android/kepler.apk 
ie http://www.orvibo.com/software/android/kepler.apk 
and it should work

**Comment by vrm42** [comment]
OMG, it's working. I mean the whole thing, not just the key extraction. You're great!
I'd definitely want to buy you a beer or two. Got any cryptocurrency address?
One more question: can I send direct ON and OFF command to the socket? Or I only can toggle its state.
Thank you for your work on this!

**Comment by sandysound** [comment]
Awesome! glad you got it all working! My sockets sometimes disconnect if I send an off command to an already off device which is why I just added a toggle but you can easily add something like this to your code to make it work with just on or off:

```
turnOn(uid) {
    let sockets = orvibo.getConnectedSocket();
    let socket = sockets.find(socket => uid === socket.uid);
    if (socket == null || socket.state === 0) {
        return;
    }
    orvibo.toggleSocket(socket.uid);
};

turnOff(uid) {
    let sockets = orvibo.getConnectedSocket();
    let socket = sockets.find(socket => uid === socket.uid);
    if (socket == null || socket.state === 1) {
        return;
    }
    orvibo.toggleSocket(socket.uid);
};
```
I don't have an cryptocurrency address but I do have a paypal link that is always happy to receive beer money haha http://paypal.me/sandysounds

**Comment by vrm42** [comment]
Well, I can't add it to my code since I don't have any :)
I'm not a coder, I only can cope with PHP a little. But it is enough for me. I've put a basic home controller together based on OrangePI with Fedora 22 minimal, USB relay boards, USB thermometer and smart sockets. I can control lights, air vents, heater, anything via cron and my smart phone's kiosk browser.
I put your http toggle url into php with curl, it's working fine. Somehow I just overlooked the part you added about the status report, but now I'm aware of that so I can control the toggling process.

Your solution is working fine so it would deserve a little more detailed guide for less experienced users like me. For example, the DNS thing could be very complicated. For me, it was relatively easy: I installed dnsmasq on OPI, I have pfSense router so I just reserved a fix IP for the socket and specified a non-existent IP as gateway and the OPI's IP as DNS server.
Four incoming ports (tcp 53,3000,10001 and udp 53) must be enabled on the controller.
Finally the server should be started automatically, like a service. I tried to make a service like I did with other stuffs before but it didn't work. I ended up with installing PM2, then
pm2 start /path/to/Example.js
pm2 save
pm2 startup

I just ordered a second B25, it's nice to see they work in my use case.
Thanks for all, tonight's beers on me :)

**Comment by sandysound** [comment]
Sounds like you've done well to setup such a system without much coding knowledge. Yes you can use the state field in the socket json objects to see if they are currently on or off. 
Thanks for the beer donation much appreciated, I'll get some extra nice craft beers with that! Glad you're getting it all up and running. Will close this issue for now ;)

**Comment by wichopy** [comment]
@vrm42 I'm trying to set this up on my local network and am stuck on the dnsmasq portion of the set up. I set a static IP for my B25 socket on my router, installed dnsmasq on ubuntu and pasted the example dnsmasq.conf code:

server=8.8.8.8
address=/homemate.orvibo.com/192.168.1.145

After running the node server, i am stil not able to intercept the packets. How do you test that you set up dns masq properly and is there any other configuration needed to get this running properly?

**Comment by vrm42** [comment]
Fortunately I recorded the neccesary steps. This is for Fedora 22, I hope you can translate it to Ubuntu. I have pretty much zero knowlege regarding to linux.

CONFIGURE B25
- install android app HOMEMATE
- press the button until it blinks red
- set wifi credentials with the app
- test on/off
- you can uninstall the app

CONFIGURE NETWORK FOR B25
- reserve fix IP in DHCP for the socket
- set gateway for this reservation to a non-existent IP or create firewall rule so socket can't access the internet
- set DNS server for this reservation to the OPI's IP
- note the MAC address of the socket

PREREQUISITES
$ yum clean all
$ yum -y update
$ yum -y install tar binutils wget curl git

INSTALLING NODE.JS
[https://www.liquidweb.com/kb/how-to-install-node-js-via-nvm-node-version-manager-on-fedora-21]
$ curl https://raw.githubusercontent.com/creationix/nvm/v0.33.4/install.sh | bash
; for latest version check https://github.com/creationix/nvm/releases
; close and re-open putty
$ nvm install stable

EXTRACTING ORVIBO COMMUNICATION KEY FROM THEIR APP
$ wget http://www.orvibo.com/software/android/kepler.apk -qO /tmp/kepler.apk
$ unzip -p /tmp/kepler.apk classes.dex | strings | grep -e '^[A-Za-n1-9]\{16\}$'
; note the key

INSTALLING DNS SERVER
$ yum -y install dnsmasq
$ systemctl enable dnsmasq
; open /etc/dnsmasq.conf
; look for (CTRL+W) 'address=' section and add 'address=/homemate.orvibo.com/192.168.1.143'
$ systemctl start dnsmasq

OPEN INCOMING PORTS
$ firewall-cmd --permanent --zone=public --add-port=3000/tcp
$ firewall-cmd --permanent --zone=public --add-port=10001/tcp
$ firewall-cmd --permanent --zone=public --add-port=53/tcp
$ firewall-cmd --permanent --zone=public --add-port=53/udp
$ firewall-cmd --reload

INSTALLING FAKE HOMEMATE SERVER
$ git clone https://github.com/sandysound/orvibo-b25-server.git
; add the key and MAC (without dashes) to the OrviboSetting.js file
$ npm install 
$ npm start

USING THE SOCKET
$ curl http://192.168.1.143:3000
; returns the states (1=off, 0=on), names and uuids (=mac addresses) of the available sockets
$ curl http://192.168.1.143:3000/?uid=5ccf7f25d9ad
; toggles the specified socket

RUN HOMEMATE SERVER AS A SYSTEM SERVICE
$ npm install pm2@latest -g
$ pm2 start /path/to/Example.js
$ pm2 save
$ pm2 startup


That's it. To check your DNS:
- set a client OS to dhcp
- reserve an IP for that client like you did it for the socket
- on your client ping homemate.orvibo.com. your ubuntu should reply.

Don't forget to open the neccesary ports for incoming connections on your ubuntu server!

**Comment by wichopy** [comment]
@vrm42 You are amazing! Even though you say you don't know much, you seem to know quite a bit about networking and linux :P

I am currently stuck on the **CONFIGURE NETWORK FOR B25** step. Maybe its my routers lack of configuration settings, but I cannot find a way of setting the gateway and dns server for a static route.

In issue  #3  they managed to stop network traffic to the socket using ifconfig. I will be researching that some more and also getting ddwrt on my router for some extra configuration options.

One last thing, you used the acronym OPI a few times, could you share what this means?

**Comment by vrm42** [comment]
Yes, most routers don't have extra options to configure DHCP as this use case requires. I use pfSense as a virtual machine on ESXi. You can run it on other platforms like Hyper-V and KVM, maybe VMWare Player. It is based on FreeBSD which afaik requires x86 or x64 CPU architect. Or probably you can try to install a DHCP server on your linux distro as it may have the feature you need.  Then use it as the only DHCP server for your network. 

OPI stands for Orange Pi - a cheap alternative to Raspberry Pi on which I run Fedora 22.

**Comment by sandysound** [comment]
Hey @wichopy how did you go with this?
@vrm42 has provided a pretty awesome list of steps to get you up and running but 
just for your info my setup is a little more simple in that I am just using my standard Netgear Wifi router, rather than pfsense or an open source router based on a linux distribution.
 
I have an ubuntu server running the node server and dnsmasq.  Lets say this is at 10.10.0.1 on my network. On this server I have the dnsmasq config like this:
```
server=8.8.8.8
address=/homemate.orvibo.com/127.0.0.1
```
This basically sends all DNS requests to google's DNS servers for resolution unless the request is for homemate.orvibo.com which it redirects to itself where the node server is running (via PM2 like @vrm42 explained)

Then my in my router's DNS settings I changed it from my ISP's default settings to 10.10.0.1 so that all the devices on my network now use the ubuntu server to resolve the DNS requests.

Once I plug in the sockets they all start connecting to the server ;)

**Comment by wichopy** [comment]
@vrm42 @sandysound Thank you both for providing clarity on setting up this awesome project. I was in the process of setting up ddwrt on an additional router in my network but was not able to finish yesterday. 

My mixup with the readme was the dnsmasq step. I did not realize you were using the ubuntu server running orbvibo node to also be the dns server for your entire network! This makes a lot more sense. I prefer this method over pfsense or DDWRT as I want to keep things as lightweight as possible. 
After work I will try pointing the dns server on my router to my raspberry pi and report back :)

**Comment by vrm42** [comment]
Yeah, like I said months ago this project deserves a detailed guide :)
For the record, I didn't make things complicated for myself with the pfsense part, it's been the only router I have for years. But indeed, if you don't have it already, you don't need to set up one.

**Comment by wichopy** [comment]
If I get this working I'll put in a pull request for the readme when I get some time later this week ;)

**Comment by wichopy** [comment]
Reporting in with updates. 

I finally got a chance to try things out with my rasbpi as the dns server. I had a few oversights that I will document here for anyone else that might have issues. 

I did not realize from the readme that the uid's were actually mac addresses. Thanks to @vrm42 's guide I pieced that together. Another small mistake I made along the way was having homemate.orvibo.com point to 127.0.0.1 which caused all other devices to point to themselves when I did a ping. 

Fixed it to point to the static ip address of my rasbpi and like magic everything is working :D Now that the networking part is done I plan on hooking it up to google voice assistant next.

Thank you both so much!
